server/src/main/java/org/elasticsearch/search/rank/feature/RankFeatureShardRequest.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.rank.feature;
+
+import org.elasticsearch.action.IndicesRequest;
+import org.elasticsearch.action.OriginalIndices;
+import org.elasticsearch.action.search.SearchShardTask;
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.search.internal.ShardSearchContextId;
+import org.elasticsearch.search.internal.ShardSearchRequest;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.transport.TransportRequest;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+
+/**
+ * Shard level request for extracting all needed feature for a global reranker
+ */
+
+public class RankFeatureShardRequest extends TransportRequest implements IndicesRequest {
+
+    private final OriginalIndices originalIndices;
+    private final ShardSearchRequest shardSearchRequest;
+
+    private final ShardSearchContextId contextId;
+
+    private final int[] docIds;
+
+    public RankFeatureShardRequest(
+        OriginalIndices originalIndices,
+        ShardSearchContextId contextId,
+        ShardSearchRequest shardSearchRequest,
+        List<Integer> docIds
+    ) {
+        this.originalIndices = originalIndices;
+        this.shardSearchRequest = shardSearchRequest;
+        this.docIds = docIds.stream().flatMapToInt(IntStream::of).toArray();
```

Comment:
I always wonder with stuff like this, why don't we just pass an `int[]` around from upstream code in the first place, why even build the boxed list?

Reply:
The specific reference comes from the `SearchPhaseController#fillDocIdsToLoad` which splits the docids based on the shard index. Since this is used in other places as well, maybe we could look into it in a separate PR?

Reply:
Ah makes sense :) not urgent or super important a follow-up or ignoring this is just fine :)